File: frontend/src/types/payment.ts
export type PaymentMethod =
  | 'MOCK_BANK_TRANSFER'
  | 'MOCK_CASH'
  | 'MOCK_E_WALLET'
  | 'MOMO'

export type PaymentStatus = 'PENDING' | 'SUCCESS' | 'FAILED'

export type Payment = {
  id: number
  billId: number
  ownerId: number
  tenantId: number
  amount: number
  method: PaymentMethod
  status: PaymentStatus
  paidAt: string | null
  createdAt: string
  updatedAt: string
}

export type MomoPayment = {
  paymentId: number
  billId: number
  amount: number
  method: 'MOMO'
  status: PaymentStatus
  partnerCode: string
  orderId: string
  requestId: string
  payUrl: string | null
  deeplink: string | null
  qrCodeUrl: string | null
  resultCode: number | null
  message: string | null
}

File: backend/src/main/java/com/example/rental/payment/Payment.java
package com.example.rental.payment;

import java.math.BigDecimal;
import java.time.Instant;

import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import com.example.rental.bill.Bill;
import com.example.rental.user.User;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "payments")
@EntityListeners(AuditingEntityListener.class)
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Payment {
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@ManyToOne(fetch = FetchType.LAZY, optional = false)
	@JoinColumn(name = "bill_id", nullable = false)
	private Bill bill;

	@ManyToOne(fetch = FetchType.LAZY, optional = false)
	@JoinColumn(name = "owner_id", nullable = false)
	private User owner;

	@ManyToOne(fetch = FetchType.LAZY, optional = false)
	@JoinColumn(name = "tenant_id", nullable = false)
	private User tenant;

	@Column(nullable = false, precision = 12, scale = 2)
	private BigDecimal amount;

	@Enumerated(EnumType.STRING)
	@Column(nullable = false)
	private PaymentMethod method;

	@Enumerated(EnumType.STRING)
	@Column(nullable = false)
	private PaymentStatus status;

	@Column
	private Instant paidAt;

	@Column(length = 200)
	private String externalOrderId;

	@Column(length = 50)
	private String externalRequestId;

	@Column(length = 1000)
	private String payUrl;

	@Column(length = 1000)
	private String deeplink;

	@Column(length = 1000)
	private String qrCodeUrl;

	private Integer gatewayResultCode;

	@Column(length = 255)
	private String gatewayMessage;

	@CreatedDate
	@Column(nullable = false, updatable = false)
	private Instant createdAt;

	@LastModifiedDate
	@Column(nullable = false)
	private Instant updatedAt;
}
